openhtf/io/frontend/src/app/test.ts

```diff
@@ -0,0 +1,191 @@
+// Copyright 2016 Google Inc.All Rights Reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+/** Represents an active or past test run. **/
+export class Test {
+  dutID: string;
+  startTime: number;
+  endTime: number;
+  duration: number;
+  outcome: string;
+  phases: Phase[];
+  logEntries: LogEntry[];
+
+  /**
+   * Create a Test.
+   * @param record - A serialized OpenHTF TestRecord.
+   */
+  constructor() {
+    this.duration = 0;
+    this.logEntries = [];
+    this.phases = [];
+  }
+
+  /**
+   * Update state based on a test record.
+   * @param test - A serialized OpenHTF TestState.
+   * @param phaseList - An option list of phases to override the record.
+   */
+  update(record: any) {
+    // Test state is tricky to track and update particularly because we want
+    // log entries and other data displayed in the gui to be selectable. If
+    // new components are being created on every update, we will lost active
+    // selections even when the gui appears to remain static. Instead, we need
+    // to keep the gui _actually_ static when no changes are happening, which
+    // means the inputs to view componts need to stay pointing at the same
+    // objects in memory. If someone has a better way to do this, please let
```

Comment:
Instead of storing N states I'm thinking of storing a starting state and N patches (or N patches and a state at t-N if you prefer to think of it that way), and returning the series of patches from #X to #Y on a request. Thus trading CPU for memory on the server. Any holes you guys see with that approach?